### DISCUSSION TITLE
Current and future usage of the dollar sign in identifiers

### DISCUSSION THREAD

**Discussion by PierreR** [cue-lang/cue#1263] [General]
Hi,

Is it safe to mark some custom abstractions (some kind of "generator" in our case) with the dollar (`$`) sign ?

I believe `$` has no meaning in CUE. Is it ok to assume it will stay that way ?

I can think of any other "marker" for an identifier.

Note that as the language evolves we might change the way we generate resources (via template for now).

Cheers

**Comment by verdverm** [comment]
I use attributes `@somename(anystring)` for denoting generators and other things. These are meant for tools that build on CUE and are not accessible in the CUE code.

https://github.com/hofstadter-io/hof/blob/_dev/hof.cue#L11

CUE does pay attention to dollar signs in some situations, like when dealing with `$ref` in JSONSchema. I think I may have seen something about wanting to do better about creating these when exporting to JSONSchema?

Do you have a larger example of what you are trying to do? CUE has a concept of "template" and I'm not sure if that is what you are referring to

**Reply by PierreR** [reply]
@verdverm  thanks for your quick reply

By "generator" I meant templating. Not sure this is the way to go but we are currently using the following pattern (simplified code):

#### 1. "Generator" abstraction

```cue
// There can be only one _stack for any cluster/stack combination
_stack: [Cluster=_]: [Stack=_]: {
	cluster:  Cluster
	stack:    Stack

        active: bool | *false
}

for _, i in _stack for _, v in i {
       // creation of a bunch of k8s resources here
      // route is just an placeholder
     // the following lines will use 'v.' to retrieve params
       route: "myroute": {}
}
```

#### 2. Client using the abstraction

```
_stack: dev: foo: {
        active: true
}
```

To mark `_stack` as something that will "generate" a bunch of resources I was thinking of `_$stack$`.

As an extra note, we are using CUE to generate yaml that will be used by Argo CD (gitOps).
We are re-generating all the resources (output) from a src directory in one go for any change (we currently don't rely on attributes to pass input)

**Reply by verdverm** [reply]
I opened https://github.com/cue-lang/cue/issues/1264 w.r.t. `$` usage within identifiers. Given that it is a valid character within a label, your usage appears to be OK and more of a style question.

There looks to be some overlap with the problems discussed here: https://github.com/cue-lang/cue/discussions/411

Do you want to evaluate and generate all clusters and stacks every time? There may be a better way to organize things to avoid this. (https://cuelang.org/docs/concepts/packages/#file-organization) There are some potential features that may allow other solutions

Will all of these resources be defined in the same repository? If not, how will you enforce the one combo constraint?

To make sure I am clear, the cluster/stack is a separate usage of CUE from the Argo/gitOps?

**Reply by PierreR** [reply]
@verdverm  Thanks for opening #411.
At the end I believe we will mark our template "abstraction" with `#lowercase` and use `#Uppercase` for real definition. I have realized that `#` makes more sense than `_` because I do want this kind of abstraction to be closed (nothing should come and change it after it is defined).

To answer your question, yes we want all resources to be defined in the same repository (and use GitOps/ArgoCD to push the new desired state to all clusters). Obviously here the main issue is going to be the performance ... If it is too big of an issue I guess we will have to organize things differently as suggested ;-)
